components/builder-depot/src/server.rs

```diff
@@ -461,32 +466,76 @@ fn download_latest_origin_encryption_key(req: &mut Request) -> IronResult<Respon
         OriginPublicEncryptionKey,
     >(req, &request) {
         Ok(key) => key,
-        Err(err) => return Ok(render_net_error(&err)),
+        Err(err) => {
+            if err.get_code() == ErrCode::ENTITY_NOT_FOUND {
+                match generate_origin_encryption_keys(&origin, req) {
+                    Ok(key) => key,
+                    Err(Error::NetError(e)) => return Ok(render_net_error(&e)),
+                    Err(_) => unreachable!(),
+                }
+            } else {
+                return Ok(render_net_error(&err));
+            }
+        }
     };
 
     let xfilename = format!("{}-{}.pub", key.get_name(), key.get_revision());
     download_content_as_file(key.get_body(), xfilename)
 }
 
-fn generate_origin_encryption_keys(req: &mut Request) -> IronResult<Response> {
-    debug!("Generate Origin Encryption Keys {:?}", req);
-    match get_param(req, "origin") {
-        Some(origin) => {
-            if !check_origin_access(req, &origin).unwrap_or(false) {
-                return Ok(Response::with(status::Forbidden));
-            }
+fn generate_origin_encryption_keys(
+    origin: &str,
+    req: &mut Request,
+) -> Result<OriginPublicEncryptionKey> {
+    debug!("Generate Origin Encryption Keys {:?} for {:?}", req, origin);
 
-            match helpers::get_origin(req, origin) {
-                Ok(origin) => {
-                    match helpers::generate_origin_encryption_keys(req, origin) {
-                        Ok(_) => Ok(Response::with(status::Created)),
-                        Err(err) => Ok(render_net_error(&err)),
-                    }
-                }
-                Err(err) => Ok(render_net_error(&err)),
+    match helpers::get_origin(req, origin) {
```

Comment:
Is `generate_origin_encryption_keys` only called in one place?  I couldn't tell definitively from this diff, but if the `download_latest_origin_encryption_key` function above is the only place that calls it, you can probably skip this call to `helpers::get_origin` (which I think makes a network trip to the db server, no?) since `download_latest_origin_encryption_key` already does that.